File: apps/api/internal/api/review_feedback_test.go
package api

import (
	"encoding/json"
	"strings"
	"testing"
	"time"

	"github.com/agentcanvas/api/internal/store"
	"github.com/google/uuid"
)

// Review feedback (TDM-161). deriveReviewFeedback is pure, so the whole rule —
// which events count as "came back", which ones have gone stale, and where the
// reason is read from — is provable without a database.

var reviewNow = time.Date(2026, 8, 1, 12, 0, 0, 0, time.UTC)

// reviewTask builds a task row with an optional audit log and error column.
func reviewTask(state string, errText string, audit ...store.ContentAudit) *store.Action {
	p := map[string]any{"title": "Wire the rework channel"}
	if len(audit) > 0 {
		p["audit"] = audit
	}
	raw, _ := json.Marshal(p)
	a := &store.Action{ID: uuid.New(), Kind: "action", Type: "task", State: state,
		Payload: raw, CreatedAt: reviewNow, UpdatedAt: reviewNow}
	if errText != "" {
		e := errText
		a.Error = &e
	}
	return a
}

func reworkEntry(actor, note, at string) store.ContentAudit {
	return store.NewStateAudit(actor, "done", "approved", note, mustTime(at))
}

func mustTime(s string) time.Time {
	t, err := time.Parse(time.RFC3339, s)
	if err != nil {
		panic(err)
	}
	return t
}

func TestDeriveReviewFeedback(t *testing.T) {
	// The reason a reviewer writes is long enough that the 80-rune audit summary
	// excerpt would cut it — which is exactly the case Note exists for, and the
	// one this channel must not repeat.
	longReason := "The rework handler never checks the claim fence, so a worker whose lease " +
		"was superseded can still bounce the task. Cover that case in the API tests before " +
		"this goes back through review."

	cases := []struct {
		name    string
		action  *store.Action
		want    *reviewFeedback
		wantNil bool
	}{
		{
			// The human gate's no, with its reason. `by` is a fact here, not a
			// guess: the reject endpoint admits nobody the server cannot see as a
			// signed-in human.
			name:   "a rejection carries its reason",
			action: reviewTask("rejected", "messaging is a separate service — do not touch it"),
			want: &reviewFeedback{
				Outcome: reviewRejected,
				Reason:  "messaging is a separate service — do not touch it",
				By:      AuthorHuman,
				At:      reviewNow.Format(time.RFC3339),
				State:   "rejected",
			},
		},
		{
			// Silence is reported as silence. Hiding a reasonless rejection would
			// leave the author with no signal at all that its ticket was cut.
			name:   "a rejection with no reason still reports",
			action: reviewTask("rejected", ""),
			want: &reviewFeedback{Outcome: reviewRejected, By: AuthorHuman,
				At: reviewNow.Format(time.RFC3339), State: "rejected"},
		},
		{
			// The reviewer's bounce (TDM-154), read from the audit note VERBATIM.
			name: "a rework bounce carries the reviewer's whole instruction",
			action: reviewTask("approved", "",
				reworkEntry("agent:codex-reviewer", longReason, "2026-08-01T11:00:00Z")),
			want: &reviewFeedback{
				Outcome: reviewRework,
				Reason:  longReason,
				By:      "agent:codex-reviewer",
				At:      "2026-08-01T11:00:00Z",
				State:   "approved",
			},
		},
		{
			// A human reopening finished work writes the SAME audit pair. It reads
			// through the same channel on purpose — to the author it is the same
			// news — and `by` is what tells the two apart.
			name: "a human reopen reads as the same event",
			action: reviewTask("approved", "",
				reworkEntry(AuthorHuman, "the migration is missing", "2026-08-01T11:30:00Z")),
			want: &reviewFeedback{Outcome: reviewRework, Reason: "the migration is missing",
				By: AuthorHuman, At: "2026-08-01T11:30:00Z", State: "approved"},
		},
		{
			// Claimed and being redone: still outstanding, because the worker
			// holding it is the one who needs the instruction.
			name: "a bounce is still outstanding while the task is executing",
			action: reviewTask("executing", "",
				reworkEntry("agent:reviewer", "add the test", "2026-08-01T11:00:00Z")),
			want: &reviewFeedback{Outcome: reviewRework, Reason: "add the test",
				By: "agent:reviewer", At: "2026-08-01T11:00:00Z", State: "executing"},
		},
		{
			// Answered. The bounce moved it to 'approved'; for it to be 'done' now
			// it must have been re-claimed and re-completed since. Re-quoting a
			// satisfied instruction would be worse than saying nothing.
			name: "a bounce that has since been redone is not reported",
			action: reviewTask("done", "",
				reworkEntry("agent:reviewer", "add the test", "2026-08-01T11:00:00Z")),
			wantNil: true,
		},
		{
			// Two bounces: the live one is the last.
			name: "the most recent bounce wins",
			action: reviewTask("approved", "",
				reworkEntry("agent:a", "first pass", "2026-08-01T09:00:00Z"),
				reworkEntry("agent:b", "second pass", "2026-08-01T11:00:00Z")),
			want: &reviewFeedback{Outcome: reviewRework, Reason: "second pass",
				By: "agent:b", At: "2026-08-01T11:00:00Z", State: "approved"},
		},
		{
			// TDM-160's reject-with-undo. Read off CURRENT state, so a withdrawn
			// rejection leaves no ghost — even though the error column still holds
			// the old text.
			name:    "an undone rejection stops being feedback",
			action:  reviewTask("proposed", "cut this"),
			wantNil: true,
		},
		{
			// Ordinary work in flight has nothing to say.
			name:    "a plain approved task has no feedback",
			action:  reviewTask("approved", ""),
			wantNil: true,
		},
		{
			// Other state moves in the log are not bounces.
			name: "an unrelated state move is not a bounce",
			action: reviewTask("approved", "",
				store.NewStateAudit(AuthorHuman, "executing", "approved", "released it", reviewNow)),
			wantNil: true,
		},
		{name: "nil action", action: nil, wantNil: true},
	}

	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			got := deriveReviewFeedback(c.action)
			if c.wantNil {
				if got != nil {
					t.Fatalf("want no feedback, got %+v", got)
				}
				return
			}
			if got == nil {
				t.Fatalf("want %+v, got nil", c.want)
			}
			if *got != *c.want {
				t.Fatalf("want %+v\ngot  %+v", *c.want, *got)
			}
		})
	}
}

// A rejection that came through the board's move path records WHO in the audit
// log; that beats the endpoint-derived default, which is the fallback.
func TestReviewFeedbackPrefersRecordedActor(t *testing.T) {
	a := reviewTask("rejected", "not a thing",
		store.NewStateAudit("anonymous", "proposed", "rejected", "not a thing",
			mustTime("2026-08-01T10:00:00Z")))
	got := deriveReviewFeedback(a)
	if got == nil {
		t.Fatal("want feedback")
	}
	if got.By != "anonymous" || got.At != "2026-08-01T10:00:00Z" {
		t.Fatalf("want the recorded actor and time, got by=%q at=%q", got.By, got.At)
	}
}

// An epic rejected as a whole answers the same way a ticket does — the batch
// read is where an orchestrator learns its plan was cut.
func TestDeriveReviewFeedbackOnAnEpic(t *testing.T) {
	raw, _ := json.Marshal(map[string]any{"title": "E9 · Messaging"})
	reason := "we are not touching messaging this quarter"
	e := &store.Action{ID: uuid.New(), Kind: "action", Type: "epic", State: "rejected",
		Payload: raw, Error: &reason, CreatedAt: reviewNow, UpdatedAt: reviewNow}
	got := deriveReviewFeedback(e)
	if got == nil || got.Outcome != reviewRejected || got.Reason != reason {
		t.Fatalf("want the epic's own rejection reason, got %+v", got)
	}
}

// Rows written before TDM-154 have no verbatim note, only the excerpted summary
// line. Falling back to it is better than reporting a bounce with no reason.
func TestReviewFeedbackFallsBackToSummary(t *testing.T) {
	a := reviewTask("approved", "", store.ContentAudit{
		At: "2026-08-01T11:00:00Z", Actor: "agent:old", Change: []string{store.StateChange},
		FromState: "done", ToState: "approved", Summary: `state: "done" → "approved" — fix the tests`,
	})
	got := deriveReviewFeedback(a)
	if got == nil || !strings.Contains(got.Reason, "fix the tests") {
		t.Fatalf("want the summary line as a fallback reason, got %+v", got)
	}
}

// The list read excerpts; the single-task read never does.
func TestExcerptReason(t *testing.T) {
	short, cut := excerptReason("  keep it whole  ", 400)
	if short != "keep it whole" || cut {
		t.Fatalf("short reasons pass through untouched, got %q cut=%v", short, cut)
	}
	long := strings.Repeat("é", 500)
	got, cut := excerptReason(long, 400)
	if !cut || len([]rune(got)) != 401 {
		t.Fatalf("want 400 runes plus the ellipsis, got %d runes cut=%v", len([]rune(got)), cut)
	}
}

// The epic rollup carries the came-back tickets with their reasons — the
// batch-level read an orchestrator polls while waiting on its approvals.
func TestEpicRollupReportsReturnedTickets(t *testing.T) {
	epicID := uuid.New()
	created := reviewNow

	rejected := rollupTask(1, "Rewrite the messaging queue", epicID.String(), "rejected", "", created, created)
	cut := "messaging is a separate service — do not touch it"
	rejected.Error = &cut

	bounced := rollupTask(2, "Add the fence test", epicID.String(), "approved", "", created, created)
	payload, _ := json.Marshal(map[string]any{
		"title": "Add the fence test", "epicId": epicID.String(),
		"audit": []store.ContentAudit{
			reworkEntry("agent:codex-reviewer", "no test covers a superseded lease", "2026-08-01T11:00:00Z"),
		},
	})
	bounced.Payload = payload

	fine := rollupTask(3, "Ship it", epicID.String(), "done", "shipped", created, created)

	body := buildEpicRollups(
		[]*store.Action{epicAction(epicID, "E21 · Slop in", "approved", created, nil)},
		[]*store.Action{rejected, bounced, fine},
	)
	if len(body.Epics) != 1 {
		t.Fatalf("want one epic, got %d", len(body.Epics))
	}
	got := body.Epics[0].Returned
	if len(got) != 2 {
		t.Fatalf("want the rejected and the bounced ticket, got %d: %+v", len(got), got)
	}
	if got[0].Outcome != reviewRejected || got[0].Reason != cut || got[0].TicketID != "TDM-1" {
		t.Fatalf("rejected line is wrong: %+v", got[0])
	}
	if got[1].Outcome != reviewRework || got[1].By != "agent:codex-reviewer" {
		t.Fatalf("rework line is wrong: %+v", got[1])
	}
	// The bounced ticket is 'approved', so it is also in the counts — the two
	// lists answer different questions and a ticket can be in both.
	if body.Epics[0].Tasks.ByState["approved"] != 1 {
		t.Fatalf("a bounced ticket stays in the approved count: %+v", body.Epics[0].Tasks)
	}
	// The finished one contributes to done[], not here.
	if len(body.Epics[0].Done) != 1 || body.Epics[0].Done[0].TicketID != "TDM-3" {
		t.Fatalf("want only the finished ticket in done[], got %+v", body.Epics[0].Done)
	}
}

// A ticket proposed on its own (no epic) is still reachable from the batch read
// — otherwise its reason would need a ticket ref nobody was ever handed.
func TestEpicRollupReportsUnepicedReturned(t *testing.T) {
	loose := rollupTask(7, "Delete the old gateway", "", "rejected", "", reviewNow, reviewNow)
	reason := "that gateway is still serving 2.3.x sessions"
	loose.Error = &reason

	body := buildEpicRollups(nil, []*store.Action{loose})
	if len(body.UnepicedReturned) != 1 {
		t.Fatalf("want the loose rejected ticket, got %+v", body.UnepicedReturned)
	}
	if body.UnepicedReturned[0].Reason != reason || body.UnepicedReturned[0].TicketID != "TDM-7" {
		t.Fatalf("wrong line: %+v", body.UnepicedReturned[0])
	}
}

// The context bundle renders the same channel as prose, with the reason whole.
func TestContextBundleRendersReviewFeedback(t *testing.T) {
	reason := "the rework handler never checks the claim fence — cover that in the API tests"
	task := reviewTask("approved", "", reworkEntry("agent:codex-reviewer", reason, "2026-08-01T11:00:00Z"))
	md := renderContextMarkdown(contextBundle{
		Canvas: &store.Canvas{Name: "Tandem planning"},
		Task:   task,
	}, reviewNow)

	if !strings.Contains(md, "### Sent back for rework") {
		t.Fatalf("want a rework section, got:\n%s", md)
	}
	if !strings.Contains(md, reason) {
		t.Fatalf("want the reason VERBATIM, got:\n%s", md)
	}
	if !strings.Contains(md, "agent:codex-reviewer") {
		t.Fatalf("want the reviewer named, got:\n%s", md)
	}

	rejectedMD := renderContextMarkdown(contextBundle{
		Canvas: &store.Canvas{Name: "Tandem planning"},
		Task:   reviewTask("rejected", "messaging is a separate service"),
	}, reviewNow)
	if !strings.Contains(rejectedMD, "### Rejected") ||
		!strings.Contains(rejectedMD, "messaging is a separate service") {
		t.Fatalf("want a rejection section with its reason, got:\n%s", rejectedMD)
	}
}

File: apps/mcp-gateway/test/review-feedback.test.ts
/**
 * TDM-161 — a rejection reason reaches the agent that proposed it.
 *
 * The DERIVATION is not tested here, because the gateway does not own it: the
 * API reduces a rejection (actions.error) and a rework bounce (the done →
 * approved audit note) into one `review` block, and review_feedback_test.go
 * pins those rules. What IS the gateway's job, and what these pin:
 *
 *   1. task_get passes the server's `review` through and turns it into the
 *      author's next move (`_review`) — with the reason VERBATIM, and different
 *      prose per outcome, because a rejection and a bounce ask for different
 *      things;
 *   2. it does so on EVERY branch of task_get — epic, no epic, rollup missing;
 *   3. the batch read carries the reasons too, so an orchestrator polling for
 *      its approvals learns WHY a ticket went away and not just that the count
 *      dropped;
 *   4. queue_next stays a ready-work call: no review block, no notifications
 *      copy. The reason reaches a worker one call later, on task_get, which the
 *      handoff already tells it to make.
 */
import { test } from "node:test";
import assert from "node:assert/strict";

import { Gateway } from "../src/gateway.js";
import { handleTool } from "../src/tools.js";
import { handleFacadeTool } from "../src/facade.js";
import { FACADE_TOOLS } from "../src/server.js";

const CODE = "TESTCODE";

function json(body: unknown, status = 200): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "content-type": "application/json" },
  });
}

type Routes = Record<string, (url: URL) => Response>;

function install(routes: Routes) {
  const paths: string[] = [];
  const real = globalThis.fetch;
  globalThis.fetch = (async (input: string | URL | Request, init?: RequestInit) => {
    const url = new URL(String(input));
    const method = init?.method ?? "GET";
    paths.push(`${method} ${url.pathname}${url.search}`);
    if (url.pathname === "/api/mcp/auth") {
      return json({ token: "t", canvasId: "canvas-1", canvasName: "C", canvasCode: CODE });
    }
    if (url.pathname === "/api/canvas/agents" && method === "POST") {
      return json({ agentId: "agent-1", name: "author" }, 201);
    }
    const hit = routes[url.pathname];
    if (hit) return hit(url);
    throw new Error(`unexpected request: ${method} ${url.pathname}`);
  }) as typeof globalThis.fetch;
  return { paths, restore: () => (globalThis.fetch = real) };
}

async function connect(): Promise<Gateway> {
  const gw = new Gateway({ apiUrl: "http://api.test" });
  await handleTool(gw, "canvas_connect", { code: CODE, role: "executor", name: "author" });
  return gw;
}

const REJECTION = {
  outcome: "rejected",
  reason: "messaging is a separate service — do not touch it",
  by: "human",
  at: "2026-08-01T10:00:00Z",
  state: "rejected",
};

const BOUNCE = {
  outcome: "rework",
  reason:
    "the rework handler never checks the claim fence, so a worker whose lease was superseded " +
    "can still bounce the task — cover that case in the API tests",
  by: "agent:codex-reviewer",
  at: "2026-08-01T11:00:00Z",
  state: "approved",
};

// ── task_get is where a returned ticket explains itself ──────────────────────

test("task_get hands back the rejection reason and what to do about it", async () => {
  const { restore } = install({
    "/api/canvas/actions/task-1": () =>
      json({
        action: { id: "task-1", ticketId: "TDM-21", state: "rejected", payload: { title: "X" } },
        linked: [],
        review: REJECTION,
      }),
  });
  try {
    const res = (await handleFacadeTool(await connect(), "task_get", { id: "task-1" })) as any;
    // The server's block passes through untouched — it is the data.
    assert.deepEqual(res.review, REJECTION);
    // VERBATIM: the reason is the correction, so nothing excerpts it.
    assert.ok(String(res._review).includes(REJECTION.reason));
    assert.match(String(res._review), /REJECTED/);
    assert.match(String(res._review), /by human/);
    // The two things that turn triage into a treadmill if they are not said.
    assert.match(String(res._review), /do not\s+re-propose/i);
    assert.match(String(res._review), /task_amend/);
  } finally {
    restore();
  }
});

test("task_get reads a rework bounce as a live instruction, not a verdict", async () => {
  const { restore } = install({
    "/api/canvas/actions/TDM-21": () =>
      json({
        action: { id: "task-1", ticketId: "TDM-21", state: "approved", payload: { title: "X" } },
        linked: [],
        review: BOUNCE,
      }),
  });
  try {
    const res = (await handleFacadeTool(await connect(), "task_get", { id: "TDM-21" })) as any;
    assert.ok(String(res._review).includes(BOUNCE.reason), "the whole instruction, not an excerpt");
    assert.match(String(res._review), /SENT BACK/);
    assert.match(String(res._review), /agent:codex-reviewer/);
    assert.match(String(res._review), /task_complete again/);
    // A bounce is the SAME ticket — re-proposing it is the failure mode here.
    assert.match(String(res._review), /same ticket/i);
    // And it is not a rejection: nothing here should tell the author to stop.
    assert.doesNotMatch(String(res._review), /REJECTED/);
  } finally {
    restore();
  }
});

test("a task with nothing to report carries no review copy at all", async () => {
  const { restore } = install({
    "/api/canvas/actions/task-1": () =>
      json({ action: { id: "task-1", state: "approved", payload: { title: "X" } }, linked: [] }),
  });
  try {
    const res = (await handleFacadeTool(await connect(), "task_get", { id: "task-1" })) as any;
    assert.equal("review" in res, false);
    assert.equal("_review" in res, false);
  } finally {
    restore();
  }
});

// The epic branch of task_get takes a different return path; the reason must not
// fall out of it. This is the branch a worker in a batch actually hits.
test("the reason survives the epic-rollup branch of task_get", async () => {
  const { restore } = install({
    "/api/canvas/actions/task-1": () =>
      json({
        action: { id: "task-1", state: "approved", payload: { title: "X", epicId: "epic-1" } },
        linked: [],
        epic: { id: "epic-1", title: "E21", state: "approved" },
        review: BOUNCE,
      }),
    "/api/canvas/epics": () =>
      json({
        epics: [
          {
            id: "epic-1",
            title: "E21",
            state: "approved",
            tasks: { total: 3, byState: { approved: 1, done: 2 } },
            done: [],
            drained: false,
            summaryNeeded: false,
            returned: [
              { ...REJECTION, id: "task-9", ticketId: "TDM-19", title: "Rewrite messaging" },
            ],
          },
        ],
      }),
  });
  try {
    const res = (await handleFacadeTool(await connect(), "task_get", { id: "task-1" })) as any;
    assert.ok(String(res._review).includes(BOUNCE.reason));
    // And the batch's own casualties ride along on the epic snapshot, because a
    // worker about to start is the reader most likely to be building the same
    // mistake the human already cut next door.
    assert.equal(res.epic.returned.length, 1);
    assert.equal(res.epic.returned[0].ticketId, "TDM-19");
  } finally {
    restore();
  }
});

// ── the batch read: what an orchestrator polling for approvals learns ────────

test("board_status names the tickets that came back", async () => {
  const { restore } = install({
    "/api/canvas/actions": (url) =>
      url.search.includes("type=epic")
        ? json({ actions: [] })
        : json({ actions: [{ id: "t1", state: "approved", payload: { title: "A" } }] }),
    "/api/canvas/epics": () =>
      json({
        epics: [
          {
            id: "epic-1",
            title: "E21",
            state: "approved",
            tasks: { total: 2, byState: { approved: 1, rejected: 1 } },
            done: [],
            drained: false,
            summaryNeeded: false,
            returned: [
              { ...REJECTION, id: "task-9", ticketId: "TDM-19", title: "Rewrite messaging" },
            ],
          },
        ],
      }),
  });
  try {
    const res = (await handleFacadeTool(await connect(), "board_status", {})) as any;
    assert.match(String(res._returned), /1 ticket\(s\) came BACK/);
    assert.match(String(res._returned), /returned/);
    // The reason itself is on the epic row, not re-quoted into the note.
    assert.equal(res.epics[0].returned[0].reason, REJECTION.reason);
  } finally {
    restore();
  }
});

// ── and queue_next stays a ready-work call ───────────────────────────────────

test("queue_next carries no review block and no notifications copy", async () => {
  const { restore } = install({
    "/api/canvas/actions": () =>
      json({
        actions: [
          {
            id: "task-1",
            ticketId: "TDM-21",
            state: "approved",
            payload: { title: "Add the fence test" },
          },
        ],
      }),
  });
  try {
    const res = (await handleFacadeTool(await connect(), "queue_next", {})) as any;
    assert.equal(res.tasks.length, 1);
    assert.equal("review" in res, false);
    assert.equal("_review" in res, false);
    assert.equal("_returned" in res, false);
    for (const t of res.tasks) {
      assert.equal("review" in t, false, "the queue projection must stay a ready-work list");
    }
  } finally {
    restore();
  }
});

// The manifest is the only documentation an agent gets: if task_get does not
// SAY it answers this, an author whose ticket vanished has no reason to call it.
test("the manifest teaches where a returned ticket explains itself", () => {
  const tool = FACADE_TOOLS.find((t) => t.name === "task_get")!;
  assert.ok(tool);
  assert.match(tool.description, /review/);
  assert.match(tool.description, /reject/i);
  assert.match(tool.description, /rework/i);
});
